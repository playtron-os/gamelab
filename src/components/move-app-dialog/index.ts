export { MoveAppDialog } from "./move-app-dialog";
