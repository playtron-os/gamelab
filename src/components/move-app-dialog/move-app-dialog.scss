.move-app-dialog-rct-component {
  &__main-modal {
    z-index: 1001;

    &__overlay {
      min-width: 300px;
      max-width: 300px;
      min-height: fit-content;
      overflow: visible;
    }
    &__content {
      display: flex;
      flex-direction: column;
      justify-content: flex-start;
      gap: 1rem;
      // padding-bottom: 3rem;
      &__field {
        display: flex;
        align-items: center;
      }
    }
  }

  &__move-in-progress-modal {
    z-index: 1002;

    &__overlay {
      width: 500px;
      min-height: fit-content;
    }
    &__content {
      display: flex;
      justify-content: center;
      align-items: center;
      span {
        text-align: center;
      }
    }
  }
}
