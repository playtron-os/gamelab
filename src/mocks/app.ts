import {
  AppDownloadStage,
  AppInformation,
  AppProvider,
  AppType,
  DownloadProgress,
  InstallConfig,
  InstalledApp,
  OwnedApp,
  PlaytronApp,
  PlaytronAppType,
  PlaytronProvider
} from "@/types";

const STEAM_OWNED_APP_MOCKS: OwnedApp[] = [
  {
    provider: AppProvider.Steam,
    name: "Apex Legends",
    id: "id1",
    provider_id: "Steam_93991",
    app_type: AppType.APP
  },
  {
    provider: AppProvider.Steam,
    name: "Counter-strike: Global Offensive",
    id: "id2",
    provider_id: "Steam_93992",
    app_type: AppType.APP
  }
];

const STEAM_DOWNLOAD_STATUS_DONE_MOCK: DownloadProgress = {
  stage: AppDownloadStage.DONE,
  progress: 100,
  error: "",
  error_code: 0,
  requires_auth: false,
  requires_2fa: false
};

const STEAM_INSTALL_CONFIG_MOCK: InstallConfig = {
  version: "1.0.0",
  install_folder: "Program Files",
  install_root: "/test/path1/root",
  install_disk: "/test/path1",
  disk_size: 100,
  download_size: 100,
  executable: "executable",
  symlink_path: "symlinkPath"
};

const STEAM_INSTALLED_APP_MOCK: InstalledApp = {
  user_id: "user_id",
  owned_app: STEAM_OWNED_APP_MOCKS[0],
  download_status: STEAM_DOWNLOAD_STATUS_DONE_MOCK,
  install_config: STEAM_INSTALL_CONFIG_MOCK,
  post_install: [],
  created_at: 0,
  updated_at: 0,
  launched_at: 0,
  interacted_at: 0
};

const STEAM_PLAYTRON_PROVIDER_MOCK: PlaytronProvider = {
  provider: AppProvider.Steam,
  providerAppId: "Steam_93991",
  lastImportedTimestamp: new Date(),
  relatedAppIds: ["id"]
};

const EPIC_PLAYTRON_PROVIDER_MOCK: PlaytronProvider = {
  provider: AppProvider.EpicGames,
  providerAppId: "EpicGames_93991",
  lastImportedTimestamp: new Date(),
  relatedAppIds: ["id"]
};

const PLAYTRON_APP_MOCKS: PlaytronApp[] = [
  {
    id: "id",
    name: "Apex Legends",
    providers: [STEAM_PLAYTRON_PROVIDER_MOCK, EPIC_PLAYTRON_PROVIDER_MOCK],
    slug: "slug",
    summary: "summary",
    description: "description",
    tags: [],
    images: [],
    publishers: [],
    developers: [],
    gameSeries: undefined,
    appType: PlaytronAppType.Game,
    releaseDate: undefined,
    comingSoon: false,
    ageRating: undefined,
    credits: undefined,
    onlyPlaytron: false,
    popularity: [],
    ratings: [],
    sales: [],
    createdAt: new Date(),
    updatedAt: new Date(),
    cmsId: 0
  },
  {
    id: "test",
    name: "Counter-strike: Global Offensive",
    providers: [STEAM_PLAYTRON_PROVIDER_MOCK, EPIC_PLAYTRON_PROVIDER_MOCK],
    slug: "slug",
    summary: "summary",
    description: "description",
    tags: [],
    images: [],
    publishers: [],
    developers: [],
    gameSeries: undefined,
    appType: PlaytronAppType.Game,
    releaseDate: undefined,
    comingSoon: false,
    ageRating: undefined,
    credits: undefined,
    onlyPlaytron: false,
    popularity: [],
    ratings: [],
    sales: [],
    createdAt: new Date(),
    updatedAt: new Date(),
    cmsId: 0
  }
];

const EPIC_OWNED_APP_MOCKS: OwnedApp[] = [
  {
    provider: AppProvider.EpicGames,
    name: "Fortnite",
    id: "xe23ft",
    provider_id: "EpicGames_Fortnite",
    app_type: AppType.APP
  }
];

export const STEAM_APP_INFORMATION_MOCKS: AppInformation[] = [
  {
    app: PLAYTRON_APP_MOCKS[0],
    installed_app: STEAM_INSTALLED_APP_MOCK,
    is_downloading: false,
    is_installing: false,
    is_launched: false,
    is_owned: true,
    is_paused: false,
    is_running: false,
    owned_apps: [STEAM_OWNED_APP_MOCKS[0], EPIC_OWNED_APP_MOCKS[0]]
  },
  {
    app: PLAYTRON_APP_MOCKS[1],
    installed_app: STEAM_INSTALLED_APP_MOCK,
    is_downloading: false,
    is_installing: false,
    is_launched: false,
    is_owned: true,
    is_paused: false,
    is_running: false,
    owned_apps: [STEAM_OWNED_APP_MOCKS[0], EPIC_OWNED_APP_MOCKS[0]]
  }
];
