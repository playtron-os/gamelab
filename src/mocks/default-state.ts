export const DEFAULT_STATE_MOCK = {
  auth: {
    userId: "userId"
  }
};
