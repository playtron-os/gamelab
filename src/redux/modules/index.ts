export * from "./auth";
export * from "./app-library";
export * from "./accounts";
export * from "./move-app-dialog";
export * from "./provider-selection-dialog";
