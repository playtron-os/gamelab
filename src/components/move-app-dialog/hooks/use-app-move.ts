import { UsePlayserveReturn } from "@/hooks";
import { closeMoveAppDialog, setMovingApp } from "@/redux/modules";
import { useAppDispatch } from "@/redux/store";
import { AppInformation } from "@/types/app-library";
import { DriveInfo } from "@/types/drive";
import { MessageType, getMessage } from "@/types/playserve/message";
import { t } from "@lingui/macro";
import { useCallback } from "react";
import { flashMessage } from "redux-flash";
export const useAppMove = (playserve: UsePlayserveReturn) => {
  const { sendMessage } = playserve;
  const dispatch = useAppDispatch();

  const sendAppMoveMessage = useCallback(
    (appsToMoveInfoArray: AppInformation[], drive: DriveInfo) => {
      dispatch(setMovingApp(true));

      appsToMoveInfoArray.forEach((appToMove, index) => {
        if (!appToMove.installed_app) {
          console.error("Unable to move, installed_app is nullish");
          closeMoveAppDialog();
          return;
        }
        const message = getMessage(MessageType.AppMove, {
          owned_app_id: appToMove.installed_app.owned_app.id,
          drive: drive.name
        });
        sendMessage(message)().then((response) => {
          const isLastUpdate = index === appsToMoveInfoArray.length - 1;
          if (response.status !== 200) {
            const message = response.body.message;
            dispatch(flashMessage(t`Failed to move app: ${message}`));
          }
          if (isLastUpdate) {
            dispatch(closeMoveAppDialog());
          }
        });
      });
    },
    [sendMessage, dispatch]
  );

  return {
    moveApps: sendAppMoveMessage
  };
};
