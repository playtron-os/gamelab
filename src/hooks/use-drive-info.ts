import { useCallback, useEffect } from "react";
import { setShowDrives, setDrives, selectDrives } from "@/redux/modules";
import { UsePlayserveReturn } from "@/hooks";
import { useAppActions, useAppSelector, useAppDispatch } from "@/redux/store";

import { DriveInfoResponseBody, MessageType, getMessage } from "@/types";

export const useDriveInfo = (playserve: UsePlayserveReturn) => {
  const { sendMessage } = playserve;
  const dispatch = useAppDispatch();

  const { setShowDrives: setShowDrivesDispatch, setDrives: setDrivesDispatch } =
    useAppActions({
      setShowDrives,
      setDrives
    });
  const drives: DriveInfoResponseBody = useAppSelector(selectDrives);

  const fetchDrives = useCallback(() => {
    const message = getMessage(MessageType.DriveInfo, {});
    sendMessage(message)()
      .then((res) => {
        if (res.status === 200 && Array.isArray(res.body)) {
          const driveNames = drives.map((drive) => drive.name);
          dispatch(setDrivesDispatch(res.body));
          res.body.forEach((drive) => {
            if (!driveNames.includes(drive.name)) {
              dispatch(
                setShowDrivesDispatch({
                  drive: drive.name,
                  enabled: true
                })
              );
            }
          });
        } else {
          console.error(res);
        }
      })
      .catch((err) => {
        console.error(err);
      });
  }, [sendMessage, setShowDrivesDispatch, setDrivesDispatch, dispatch, drives]);

  useEffect(() => {
    fetchDrives();
  }, []);

  return {
    drives,
    fetchDrives
  };
};
