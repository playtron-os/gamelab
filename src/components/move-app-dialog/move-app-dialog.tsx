import React, { useCallback, useMemo, useState } from "react";
import { Button, Modal, ProgressBar } from "@playtron/styleguide";

import { Trans, t } from "@lingui/macro";

import { useDriveInfo } from "@/hooks/use-drive-info";
import { useAppDispatch, useAppSelector } from "@/redux/store";
import {
  MoveAppDialogState,
  closeMoveAppDialog,
  selectMoveAppDialogState
} from "@/redux/modules/move-app-dialog/move-app-dialog-slice";
import { usePlayserve } from "@/hooks";
import { useAppMove } from "./hooks/use-app-move";
import { DriveInfo, MessageType, AppInformation } from "@/types";
import { getDiskSize } from "@/utils/app-info";
const getAppDrive = (appInfo: AppInformation, drives: DriveInfo[]) => {
  if (!appInfo?.installed_app || !drives.length) {
    return null;
  }
  const currentPath = appInfo.installed_app.install_config.install_root;
  let biggestPathDrive = null;

  for (const drive of drives) {
    if (
      currentPath.startsWith(drive.path) &&
      (!biggestPathDrive || drive.path.length > biggestPathDrive.path.length)
    ) {
      biggestPathDrive = drive;
    }
  }

  return biggestPathDrive;
};

export const MoveAppDialog: React.FC = () => {
  const playserve = usePlayserve();
  const [progress, setProgress] = useState(0);
  usePlayserve({
    onMessage: (message) => {
      if (
        message.message_type === MessageType.AppMoveProgress &&
        message.status === 200
      ) {
        setProgress(message.body.progress);
      }
    }
  });

  const { drives } = useDriveInfo(playserve);
  const { moveApps } = useAppMove(playserve);

  const { isMoveAppDialogOpen, appInfoArray, isMovingApp } =
    useAppSelector<MoveAppDialogState>(selectMoveAppDialogState);
  const dispatch = useAppDispatch();
  const onCloseMoveAppDialog = useCallback(
    () => dispatch(closeMoveAppDialog()),
    [dispatch]
  );

  const appDrive = useMemo(
    () => getAppDrive(appInfoArray?.[0], drives),
    [appInfoArray, drives]
  );

  if (!appDrive) {
    return null;
  }

  const appName = appInfoArray?.[0].app.name;
  const modalTitle =
    appInfoArray.length === 1 ? t`Move ${appName}` : t`Move selected apps`;

  return (
    <div>
      <Modal
        className="-z-40 p-6 max-w-[465px]"
        title={modalTitle}
        isOpen={isMoveAppDialogOpen && !isMovingApp}
      >
        <div className="py-2" data-testid="move-app-dialog-modal-content">
          {appInfoArray.length === 1 && (
            <>
              <div>
                <span className="font-bold text-sm text-gray-400">{t`From:`}</span>
              </div>
              <div className="mb-2">
                <span className="font-bold text-lg">{appDrive.model}</span>
                <br />
                <span className="text-xs font-bold text-gray-400">
                  {appDrive.name}
                </span>
              </div>
            </>
          )}
          <div className="font-bold text-sm text-gray-400">
            <Trans>To: </Trans>
          </div>
          <div className="flex gap-4 items-center">
            <div className="flex flex-col gap-4 py-4">
              {drives
                .filter(
                  (drive) =>
                    !drive.needs_formatting && drive.path !== appDrive.path
                )
                .map((drive) => (
                  <div
                    key={drive.path}
                    className={`flex items-center gap-2 p-2 rounded-md cursor-pointer outline-gray-700 outline-2 outline-double hover:outline-white`}
                    onClick={() => {
                      moveApps(appInfoArray, drive);
                    }}
                  >
                    <div>
                      <span className="font-bold text-lg">{drive.model}</span>
                      <span className="font-sm text-gray-400"></span>
                      <br />
                      <div className="text-xs font-bold text-gray-400">
                        {drive.name}: {getDiskSize(drive.available_space, 0)}/
                        {getDiskSize(drive.max_size, 0)}
                      </div>
                      <div className="w-[400px]">
                        <ProgressBar
                          value={(drive.available_space / drive.max_size) * 100}
                          size="sm"
                        />
                      </div>
                    </div>
                  </div>
                ))}
            </div>
          </div>
        </div>
        <div className="flex justify-end gap-2">
          <Button
            label={t`Close`}
            disabled={isMovingApp}
            onClick={onCloseMoveAppDialog}
          />
        </div>
      </Modal>
      <Modal className="z-50 p-6" isOpen={isMovingApp}>
        <span data-testid="move-in-progress-text">
          {" "}
          <ProgressBar
            value={progress}
            label={progress ? progress + "%" : ""}
          />{" "}
          <Trans> Move In Progress... </Trans>{" "}
        </span>
      </Modal>
    </div>
  );
};
