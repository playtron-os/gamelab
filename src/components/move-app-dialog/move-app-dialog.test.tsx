import React from "react";
import { authSlice, moveAppDialogSlice } from "@/redux/modules";
import { renderWithProviders } from "@/utils";
import { MoveAppDialog } from "./move-app-dialog";
import { useDriveInfo } from "../../hooks/use-drive-info";
import { DriveInfo } from "@/types";
import { configureStore } from "@reduxjs/toolkit";
import { STEAM_APP_INFORMATION_MOCKS } from "@/mocks/app";
import { DEFAULT_STATE_MOCK } from "@/mocks/default-state";

const MOCK_DRIVES = [
  {
    name: "/test/drive1",
    path: "/test/path1",
    available_space: 1000,
    max_size: 1000
  },
  {
    name: "/test/drive2",
    path: "/test/path2",
    available_space: 1000,
    max_size: 1000
  }
] as DriveInfo[];

// Mock hooks and redux functionalities
jest.mock("@/hooks/use-drive-info", () => ({
  useDriveInfo: jest.fn().mockReturnValue({ drives: [] })
}));

const MOVE_APP_MOCK_FN = jest.fn();
jest.mock("./hooks/use-app-move", () => ({
  useAppMove: jest.fn(() => ({ moveApp: MOVE_APP_MOCK_FN }))
}));

const getMockStore = (open = false, moving = false) => {
  return configureStore({
    reducer: {
      moveAppDialog: moveAppDialogSlice.reducer,
      auth: authSlice.reducer
    },
    preloadedState: {
      ...DEFAULT_STATE_MOCK,
      moveAppDialog: {
        isMoveAppDialogOpen: open,
        appInfoArray: [STEAM_APP_INFORMATION_MOCKS[0]],
        isMovingApp: moving
      }
    }
  });
};

describe("MoveAppDialog Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders MoveAppDialog component and matches null snapshot", () => {
    const { container, asFragment } = renderWithProviders(
      <MoveAppDialog />,
      undefined,
      getMockStore()
    );

    expect(container).toBeEmptyDOMElement();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders MoveAppDialog component and matches snapshot", () => {
    (useDriveInfo as jest.Mock).mockReturnValue({
      drives: MOCK_DRIVES
    });
    const { container, asFragment } = renderWithProviders(
      <MoveAppDialog />,
      undefined,
      getMockStore(true, false)
    );

    expect(container).not.toBeEmptyDOMElement();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders MoveAppDialog modal content", () => {
    (useDriveInfo as jest.Mock).mockReturnValue({
      drives: MOCK_DRIVES
    });
    const { getByTestId } = renderWithProviders(
      <MoveAppDialog />,
      undefined,
      getMockStore(true, false)
    );

    expect(getByTestId("move-app-dialog-modal-content")).toBeInTheDocument();
  });

  it("shows move progress indicator", () => {
    (useDriveInfo as jest.Mock).mockReturnValue({
      drives: MOCK_DRIVES
    });
    const { getByTestId } = renderWithProviders(
      <MoveAppDialog />,
      undefined,
      getMockStore(true, true)
    );
    expect(getByTestId("move-in-progress-text")).toBeInTheDocument();
  });
});
