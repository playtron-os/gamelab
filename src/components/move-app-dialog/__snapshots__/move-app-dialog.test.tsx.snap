// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoveAppDialog Component renders MoveAppDialog component and matches null snapshot 1`] = `<DocumentFragment />`;

exports[`MoveAppDialog Component renders MoveAppDialog component and matches snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
